core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java

```diff
@@ -0,0 +1,362 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.admin.indices.shards;
+
+import com.carrotsearch.hppc.cursors.IntObjectCursor;
+import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
+import com.google.common.collect.ImmutableList;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.ShardOperationFailedException;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.collect.ImmutableOpenIntMap;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Streamable;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentBuilderString;
+import org.elasticsearch.rest.RestStatus;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.ExceptionsHelper.detailedMessage;
+
+/**
+ * Consists of {@link ShardStoreStatus}s for request indices grouped by
+ * indices and shard ids and a list of node {@link Failure}s encountered
+ */
+public class IndicesShardsStoresResponse extends ActionResponse implements ToXContent {
+
+    /**
+     * Shard store information from a node
+     */
+    public static class ShardStoreStatus implements Streamable, ToXContent {
+        private DiscoveryNode node;
+        private long version;
+        private Throwable storeException;
+
+        private ShardStoreStatus() {
+        }
+
+        ShardStoreStatus(DiscoveryNode node, long version, Throwable storeException) {
+            this.node = node;
+            this.version = version;
+            this.storeException = storeException;
+        }
+
+        /**
+         * node the shard belongs to
+         */
+        public DiscoveryNode getNode() {
+            return node;
+        }
+
+        /**
+         * version of the shard, the highest
```

Comment:
this is not really true - the version is used to select the store that will be used as the very first primary. The version is not used in the choice of promoting an existing replica to a primary